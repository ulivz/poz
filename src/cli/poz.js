import { errorListLogger, localPackagesLogger } from './utils'
// import { exec, execSync } from '../utils'
import { POZ, PackageManager } from '../index'
import shelljs from 'shelljs'

export default function (cli) {
  return {
    command: {
      name: '*',
      opts: {
        desc: 'Programmable scaffolding generator',
      },
      handler(input, flags) {
        const pm = new PackageManager()

        if (!input.length) {
          cli.showHelp()
          localPackagesLogger(pm.cache.getItem('packagesMap'))

        } else {
          const requestName = input[0]
          const TIMEOUT = 60000
          const prevCwd = process.cwd()
          pm.fetchPkg(requestName, TIMEOUT)
            .then(pkg => {
              if (pkg) {
                process.chdir(pkg.cachePath)
                shelljs.exec(`yarn`)
                process.chdir(prevCwd)
                const app = new POZ(pkg.cachePath)
                return app.launch()
              }
            })
            .catch(error => {
              if (error.length && error[0] instanceof Error) {
                errorListLogger(requestName, error)
              } else {
                console.log(error)
              }
            })
        }
      }
    },
    options: [
      {
        name: 'local',
        opts: {
          alias: '-l',
          desc: 'Use a local poz package',
          type: 'boolean'
        }
      }
    ]
  }
}
