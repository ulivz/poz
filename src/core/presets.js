export const LIFE_CYCLE = [
  'onStart',
  'onPromptStart',
  'onPromptEnd',
  'onDestStart',
  'onDestEnd',
  'onExit'
]

export const TEMPLATE_DIRECTORY_NAME = 'template'
export const PACKAGE_ENTRY_FILE_NAME = 'poz.js'

