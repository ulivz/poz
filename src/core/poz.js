import alphax from 'alphax'
import { relative } from 'path'
import { getNormalizedConfig } from './normalize-config'
import { LIFE_CYCLE } from './presets'
import { RENDER_FAILURE, RENDER_SUCCESS } from './event-names'
import { isDev, isTest } from './env'
import * as env from './env'
import event from './event'
import Context from './context.js'
import validatePackage  from './package-validator'
import saoPlugin  from '../plugins/poz-sao-plugin'

import utils, { logger as log, prompts, datatypes, consolelog, assert, getGitUser } from '../utils/index'

const { promptsRunner, mockPromptsRunner, promptsTransformer } = prompts
const { isFunction } = datatypes

class POZ {
  constructor(src, { write = true } = {}) {
    this.src = src
    this.templateDir = null
    this.write = write
    this.userConfig = null
    this.normalizedConfig = null
    this.user = {}
    this.status = null
    this.app = null
    this.error = null
    this.plugins = []
    this.cwd = process.cwd()
    this.context = Context()
    this.entryFileName = ['poz.js']
    this.templateDirName = 'template'
    if (arguments[1] === false) {
      this.write = false
    }
    this.use(saoPlugin)
    return this
  }

  use(plugin) {
    this.plugins.push(plugin)
  }

  addEntryFileName(name) {
    this.entryFileName.push(name)
  }

  validatePackage() {
    const {
      error,
      templateDir,
      userConfig
    } = validatePackage(this.src, {
      entryFileName: this.entryFileName,
      templateDirName: this.templateDirName,
      exportArguements: [this.context, utils]
    })
    this.error = error
    this.templateDir = templateDir
    this.userConfig = userConfig
    return this
  }

  initializeContext() {
    this.user = getGitUser()
    this.context.assign({
      cwd: this.cwd,
      env: env,
      dirname: this.cwd.slice(this.cwd.lastIndexOf('/') + 1),
      gituser: this.user.name,
      gitemail: this.user.email,
      src: this.src
    })
  }

  bindHookListener() {
    for (const hookname of LIFE_CYCLE) {
      const handler = this.userConfig[hookname]
      if (!handler) {
        continue
      }
      if (isFunction(handler)) {
        event.on(hookname, (...args) => {
          consolelog(isTest, `Running ${log.cyan(hookname)} ...`)
          this.status = hookname
          handler(...args)
        })
      } else {
        consolelog(isTest, `Expect ${log.cyan(hookname)} to be a function`)
      }
    }
  }

  handleRenderSuccess(file) {
    let filePath = relative(this.src, file.path)
    log.success(`render ${log.cyan(filePath)}`)
  }

  handleRenderFailure(error, file) {
    let filePath = relative(this.src, file.path)
    log.error(`render ${log.cyan(filePath)}`)
    log.echo(error)
  }

  normalizeUserConfig() {
    this.normalizedConfig = getNormalizedConfig(this.userConfig, this.context)
    this.context.set('config', this.normalizedConfig)
  }

  prompt() {
    event.emit('onPromptStart')
    let { prompts } = this.userConfig
    const promptsMetadata = isFunction(prompts) ? prompts() : prompts
    prompts = promptsTransformer(promptsMetadata)
    const envPromptsRunner = isTest
      ? mockPromptsRunner
      : promptsRunner
    return envPromptsRunner(prompts)
  }

  dest() {
    this.app = alphax()
    const { rename, filters, render, outDir } = this.normalizedConfig
    this.app.src(this.templateDir + '/**', {
      baseDir: this.templateDir,
      rename,
      filters,
      transform: render
    })
    return this.app.dest(outDir || '.', { write: this.write })
      .then(() => this.app.fileMap())
      .catch(error => {
        assert(isTest, error)
      })
  }

  launch() {
    this.plugins.forEach(plugin => plugin(this))
    this.validatePackage()

    event.on(RENDER_FAILURE, this.handleRenderFailure.bind(this))
    event.on(RENDER_SUCCESS, this.handleRenderSuccess.bind(this))

    this.error.throwIfError()
    this.initializeContext()
    this.bindHookListener()

    event.emit('onStart')
    return this.prompt()

      .then(answers => {
        this.context.assign(answers)
        event.emit('onPromptEnd')
        this.normalizeUserConfig()
        event.emit('onDestStart')
        return this.dest()
      })

      .then(files => {
        event.emit('onDestEnd')
        event.emit('onExit')
        return files
      })

      .catch(error => {
        throw error
        event.emit('onExit', error)
      })
  }
}

export default (...args) => new POZ(...args)
