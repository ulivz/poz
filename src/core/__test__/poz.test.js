import poz from '../poz'
import { LIFE_CYCLE } from '../presets'
import { resolve } from 'path'

describe('POZ', () => {

  let prevCwd = process.cwd()

  beforeAll(() => {
    process.chdir(resolve(__dirname, 'fixtures'))
  })

  afterAll(() => {
    process.chdir(prevCwd)
  })

  function getPkgPath(dirname) {
    return __dirname + '/fixtures/POZTestPackage_' + dirname
  }

  test('Should each life cycle works', async () => {
    const pkgPath = getPkgPath('VALID')
    const app = poz(pkgPath, false)
    const files = await app.launch()
    expect(files).toMatchSnapshot()
    expect(app.context.calledLifeCycle).toEqual(LIFE_CYCLE)
  })

  test('Should throw error when giving a nonexisitng package directory', async () => {
    const pkgPath = getPkgPath('INVALID_0')
    const app = poz(pkgPath, false)
    app.validatePackage()
    const { error } = app
    expect(error.currentErrorCode()).toBe('NOT_FOUND')
  })

  test('Should throw error when not found "poz.js"', async () => {
    const pkgPath = getPkgPath('INVALID_1')
    const app = poz(pkgPath, false)
    app.validatePackage()
    const { error } = app
    expect(error.currentErrorCode()).toBe('MISSING_ENTRY_FILE')
  })

  test('Should throw error when "poz.js" is a empty JavaScript file', async () => {
    const pkgPath = getPkgPath('INVALID_2')
    const app = poz(pkgPath, false)
    app.validatePackage()
    const { error } = app
    expect(error.currentErrorCode()).toBe('CANNOT_EXPORT_EMPTY_OBJECT')
  })

  test('Should throw error when "poz.js" not export plain object or function', async () => {
    const pkgPath = getPkgPath('INVALID_3')
    const app = poz(pkgPath, false)
    app.validatePackage()
    const { error } = app
    expect(error.currentErrorCode()).toBe('UNEXPECTED_ENTRY_FILE')
  })

  test('Should throw error when missing "template" directory', async () => {
    const pkgPath = getPkgPath('INVALID_4')
    const app = poz(pkgPath, false)
    app.validatePackage()
    const { error } = app
    expect(error.currentErrorCode()).toBe('MISSING_TEMPLATE_DIRECTORY')
  })

})
