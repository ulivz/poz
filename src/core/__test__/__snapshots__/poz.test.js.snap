// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POZ Should each life cycle works 1`] = `
Object {
  "README.md": "# fixtures

- description: my awesome project
- author: POZ
",
  "src/index.js": "module.exports = function fixtures() {
  // Go hacking!
}",
}
`;
